db/indices.c

```diff
@@ -15,17 +15,146 @@
  */
 
 #include <stdio.h>
-#include <string.h>
-#include <stdlib.h>
-#include <inttypes.h>
-#include "comdb2.h"
-#include <schemachange.h>
-#include "block_internal.h"
 #include <assert.h>
+#include "schemachange.h"
+#include "block_internal.h"
 #include "logmsg.h"
 #include "indices.h"
 
+#define DEBUG_REORDER 0
+
 extern int gbl_partial_indexes;
+extern int gbl_reorder_idx_writes;
+static __thread void *defered_index_tbl = NULL;
+static __thread void *defered_index_tbl_cursor = NULL;
+
+
+//
+//del needs to sort before adds because dels used to happen online
+
+// defered index table types
+// the _CC types signify that we need to check constraints
+// for that key on the parent table
+typedef enum { DIT_DEL, DIT_UPD, DIT_ADD, DIT_ADD_CC, DIT_DEL_CC, DIT_UPD_CC} dit_t;
+
+//defered index table
+typedef struct {
+    struct dbtable *usedb; //consider not storing usedb and processing each usedb separately
+    short ixnum;
+    char ixkey[MAXKEYLEN]; //consider storing up to the largest key
+    dit_t type;
+    unsigned long long genid;
+    unsigned long long newgenid; // new genid used for update
+} dtikey_t;
+
+#define CMP_KEY_MEMBER(k1, k2, var)                                            \
+    if (k1->var < k2->var) {                                                   \
+        return -1;                                                             \
+    }                                                                          \
+    if (k1->var > k2->var) {                                                   \
+        return 1;                                                              \
+    }
+
+#define MEMCMP_KEY_MEMBER(k1, k2, var, len)                                    \
+    int __rc = memcmp(k1->var, k2->var, len);                                  \
+    if (__rc != 0 ) {                                                          \
+        return __rc;                                                           \
+    }                                                                          \
+
+
+static int defered_index_key_cmp(void *usermem, int key1len,
+                                 const void *key1, int key2len,
+                                 const void *key2)
+{
+    assert(sizeof(dtikey_t) == key1len);
+    assert(sizeof(dtikey_t) == key2len);
+ 
+    dtikey_t *k1 = (dtikey_t *)key1;
+    dtikey_t *k2 = (dtikey_t *)key2;
+
+    CMP_KEY_MEMBER(k1, k2, usedb);
+    CMP_KEY_MEMBER(k1, k2, ixnum);
+    MEMCMP_KEY_MEMBER(k1, k2, ixkey, sizeof(k1->ixkey));
+    CMP_KEY_MEMBER(k1, k2, type);
+    CMP_KEY_MEMBER(k1, k2, genid);
```

Comment:
should genid be first?

Reply:
Adding genid to end of ixkey when ix is dup. Will try how that works.